darts/ad/scorers/kmeans_scorer.py

```diff
@@ -0,0 +1,216 @@
+
+        If `component_wise` is set to True, the algorithm will be applied to each dimension independently. The distance
+        will be computed between the vector and the closest centroid found by the model trained on the corresponding
+        dimension during the training.
+
+        Parameters
+        ----------
+        window
+            Size of the window used to create the subsequences of the series.
+        k
+            The number of clusters to form as well as the number of centroids to generate by the KMeans model.
+        diff_fn
+            Optionally, reduced function to use if two series are given. It will transform the two series into one.
+            This allows the KMeansScorer to apply KMeans on the original series or on its residuals (difference between
+            the prediction and the original series).
+            Must be one of "abs_diff" and "diff" (defined in ``_diff_series()``).
+            Default: "abs_diff"
+        component_wise
+            Boolean value indicating if the score needs to be computed for each width/dimension independently (True)
+            or by concatenating the width in the considered window to compute one score (False).
+            Default: False
+        """
+
+        raise_if_not(
+            type(component_wise) is bool,
+            f"'component_wise' must be Boolean, found type: {type(component_wise)}",
+        )
+        self.component_wise = component_wise
+
+        self.k = k
+
+        if component_wise:
+            returns_UTS = False
+        else:
+            returns_UTS = True
+
+        super().__init__(returns_UTS=returns_UTS, window=window, diff_fn=diff_fn)
+
+    def __str__(self):
+        return "KMeansScorer"
+
+    def _fit_core(
+        self,
+        list_series: Sequence[TimeSeries],
+    ):
+
+        list_np_series = [series.all_values(copy=False) for series in list_series]
+
+        if not self.component_wise:
+            self.model = KMeans(n_clusters=self.k)
+            self.model.fit(
+                np.concatenate(
+                    [
+                        np.array(
+                            [
+                                np.array(np_series[i : i + self.window])
+                                for i in range(len(np_series) - self.window + 1)
+                            ]
+                        ).reshape(-1, self.window * len(np_series[0]))
+                        for np_series in list_np_series
+                    ]
+                )
+            )
+        else:
+            models = []
+            for width in range(self.width_trained_on):
+                model = KMeans(n_clusters=self.k)
+                model.fit(
+                    np.concatenate(
+                        [
+                            np.array(
+                                [
+                                    np.array(np_series[i : i + self.window, width])
```

Comment:
You should definitely try to find way to vectorize this. We can leave it like that for now in this PR, but this should be high prio for later :) Numpy offers ways to perform windowing. Our own `TimeSeries` also does. You should try to rely on those.

In general having for loops over either timestamps or dimensions (width) should be a red flag!

Reply:
Agree. It is actually the same code used in:
- Pyod Scorer
- Kmeans Scorer 
- Wasserstein Scorer 

for fitting() and scoring(). So one solution could be used 6 times!